src/renderer/mesh_storage.zig

```diff
@@ -967,7 +971,153 @@ pub fn updateLightMap(map: *LightMap.LightMapFragment) void {
+	damage: std.AutoHashMapUnmanaged(Vec3i, f32),
+	mutex: std.Thread.Mutex,
+	secondsSinceLastUpdate: f64,
+
+	const updateIntervalSeconds: f64 = 1.0;
+
+	const Update = union(enum) {
+		add: struct {pos: Vec3i, progress: f32},
+		remove: struct {pos: Vec3i},
+	};
+
+	fn init() BlockDamage {
+		return .{
+			.updateQueue = .init(main.globalAllocator, 16),
+			.damage = .{},
+			.mutex = .{},
+			.secondsSinceLastUpdate = 0,
+		};
+	}
+	fn deinit(self: *BlockDamage) void {
+		self.mutex.lock();
+		defer self.mutex.unlock();
+
+		self.updateQueue.deinit();
+		self.damage.deinit(main.globalAllocator.allocator);
+	}
+	fn clear(self: *BlockDamage) void {
+		main.utils.assertLocked(&self.mutex);
+
+		var iterator = self.damage.iterator();
+		while(iterator.next()) |entry| {
+			const pos = entry.key_ptr.*;
+			self.removeAnimation(pos);
+		}
+		self.damage.clearRetainingCapacity();
+	}
+	pub fn set(self: *BlockDamage, pos: Vec3i, remainingHealth: f32) void {
+		main.utils.assertLocked(&self.mutex);
+
+		if(remainingHealth <= 0) {
+			if(self.damage.remove(pos)) {
+				self.removeAnimation(pos);
+			}
+			return;
+		}
+
+		const block = getBlock(pos[0], pos[1], pos[2]) orelse blocks.Block.Air;
+		const maxBlockHealth = block.blockHealth();
+
+		if(remainingHealth >= maxBlockHealth) {
+			if(self.damage.remove(pos)) {
+				self.removeAnimation(pos);
+			}
+			return;
+		}
+
+		const result = self.damage.getOrPut(main.globalAllocator.allocator, pos) catch unreachable;
+		if(result.found_existing) {
+			if(result.value_ptr.* == remainingHealth) return;
+			self.removeAnimation(pos);
+		}
+		result.value_ptr.* = remainingHealth;
+		const progress = @min(1.0, @max(0.0, 1.0 - (remainingHealth/maxBlockHealth)));
+		self.addAnimation(pos, progress);
+	}
+	fn removeAnimation(self: *BlockDamage, pos: Vec3i) void {
+		self.updateQueue.enqueue(.{.remove = .{.pos = pos}});
+	}
+	fn addAnimation(self: *BlockDamage, pos: Vec3i, progress: f32) void {
+		self.updateQueue.enqueue(.{.add = .{.pos = pos, .progress = progress}});
+	}
+	pub fn get(self: *BlockDamage, pos: Vec3i) ?f32 {
+		main.utils.assertLocked(&self.mutex);
+		return self.damage.get(pos);
+	}
+	pub fn update(self: *BlockDamage, deltaTime: f64) void {
+		self.mutex.lock();
+		defer self.mutex.unlock();
+
+		self.secondsSinceLastUpdate += deltaTime;
+		while(self.secondsSinceLastUpdate > updateIntervalSeconds) {
+			self._update(updateIntervalSeconds);
+			self.secondsSinceLastUpdate -= updateIntervalSeconds;
+		}
+	}
+	fn _update(self: *BlockDamage, deltaTime: f64) void {
+		var keysToRemove: main.ListUnmanaged(Vec3i) = .{};
+		defer keysToRemove.deinit(main.stackAllocator);
+
+		var iterator = self.damage.iterator();
+		while(iterator.next()) |entry| {
+			const pos = entry.key_ptr.*;
+			const remainingHealth = entry.value_ptr.*;
+
+			const block = getBlock(pos[0], pos[1], pos[2]) orelse blocks.Block.Air;
+			const maxBlockHealth = block.blockHealth();
+
+			if(maxBlockHealth == 0) {
```

Comment:
Why is this needed? Shouldn't you explictly removing all animation states after the block is broken anyways?

Reply:
If this ever gets triggered it means that the breaking of the block wasn't synchronized correctly with the client-side breaking data structure.
That seems like a bug to me, so you should use an assertion instead.